### PR TITLE
Feature : situation report creation screen implementation

### PR DESCRIPTION
**Pull request by AdrienDeschenaux**
Hello Team! 

Here's my PR on the task of implementing the situation report creation screen, 
This PR will be structured the following : 
- Utils function : I had to implement new functions that I added to the same util file I created last week
- Screen : 
   - Screen logic/flow : A quick overview of the screen and its logic
   - Sub-component : Quick look on some local components
- Refactoring
- Prettier
- Next step

## Utils functions

Those functions were created in order to improve code readability in the front-end and split as much as possible the logic from the front-end

### `addSingleItemToGroup`

I created this function because I figured that I previously created a function that was adding a new group (or single group element) to the layout, but we weren't able to add a single element in a group, this function is handling this addition properly. 

### `removeItemFrom` 

Similarly as above, I already created last weeks a way to delete a whole group but I did not add a way to delete a single element inside of a group, which is what is done by this function.

### Screen Overview

Here's a brief overview of what is happening :

<div style="display: flex; justify-content: space-between;">
  Screenshot_20241205-190935_Expo Go
  Screenshot_20241205-191104_Expo Go
  Screenshot_20241205-191110_Expo Go
  Screenshot_20241205-191135_Expo Go
</div>

### Screen Logic

First when arriving on the screen there is a picker that allow the user to select a residence, the dimensions of the picker are, just like it was last week constrained by the fact that the component would cut the text input in it, which wouldn't be a good UI option in my opinion. The landlord can select in which residence he would like to add a situation report with the picker, and set a name for this precise layout.

Then the user is invited to tap on the edit button, which will switch to editing mode. the landlord can now edit (add/delete) new elements and change the name of those elements as well. Then tapping on save will save the created layout locally, he will then have a preview of how the layout will be displayed. Then if he likes the layout, he can scroll all the way down of the screen and finally tap on the submit button that will upload the situation report to the corresponding residence.

In order to code those functionalities, I used conditional display of the elements with a state : editMode, which made it possible for me to display input text fields and the addition button when editing and the final render when not on editing.

And for the "save"/"cancel" option, I used two different states so that while editing only the a temporary layout was modified and the original layout was not set as the temporary layout unless the user tap on the save button. An Issue I had was that I used `setTempLayout` on `layout` and `setLayout` on `tempLayout` in order to modify the states, but in fact this was causing unexpected behavior when modifying the names using the input text fields, so I switched to a deep copy of the layout which solved the problem. We can expect the layout to not be too long and so the time complexity of those deep copies should not be problematic 

## Sub-components

What I call sub-components are the local components I created in order to achieve this, a few of them are pretty similar to those of the last week, but I had to slightly modify them because there were subtle differences in the functionalities needed in this weeks screen and last weeks, I think that the best option was to retake and modify what I needed out of the original components.

Here I will briefly explain what those components are about : 

- `RemoveSingleInButton` : this component represents the button to remove the single elements it uses the function I created in util which was mentioned above
- `RemoveGroupButtonProps` : This component is in charge of properly removing a group from the layout, this removing occurs only when the group is a singleton
- `SituationReportItem` : this represents a situation report item, there is a conditional representation that allow be to have both the remove button and the input text field displayed in case the user is in edit mode
- `GroupedSituationReport` : This allow us to display an input text field when on editing mode, the addition buttons, and properly handles the display of the situation report with respect to the number of items in the group (i.e : if there is only one element then it only display the element in a gray box, if there are multiple elements then display those in a purple group) 
- `AddItemButton` : Provide and handle all the functionalities in order to be able to add a new group (either a singleton or a group containing at least 2 elements) 

## Refactoring 

I did a small refactoring on the SubmitButton component because the fact that it was wrapped in a view made it painful to use in some situations, I modified the code in the corresponding screens in order to avoid breaking the layout

I also slightly refactor the styling for the situationReportScreen, which was done last week, moving the possible styling in a separated file stored in the styles folder

## Prettier 

I installed in the project `prettier` which is a library that allow auto indentation and formatting of the code. 
In order to be able to run it you should only need to run `npx prettier --write path-to-file`.

## Next step 

The next possible improvement for this functionality could be to make the landlord able to modify an already existing layout. We could also think of allowing him to re-organize the situation report items in the layout as he likes or even to create groups by sliding single items together, this could be implemented if we have the time. I think that a bigger priority should be to make the residences creation and linking it to the landlord work as soon as possible as this core feature is not yet implemented which is too bad in my opinion because this create holes in the flow that would be really cool otherwise

I know this is not a small PR, let me know if you need clarifications on my work 
Thank you for your time !

### CODE REVIEW

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+            )}
+
+            <View
+              style={[
+                appStyles.submitContainer,
+                situationReportStyles.submitMargin,
+              ]}
+            >
+              <SubmitButton
+                disabled={layout.length === 0 || situationReportName === ''}
```
Comment (mdesautel) [inline app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:598]: I think you should also disable the button if the residence hasn't been selected yet. The useState for the residence has ' ' as default value, which is not a valid residence, and if we don't click on the picker but fill in the other things, it would be a problem I think, even if it doesn't create an immediate error.

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+  const [editMode, setEditMode] = useState(false);
+  const [tempLayout, setTempLayout] = useState<[string, [string, number][]][]>([],);
+  const [selectedResidence, setSelectedResidence] = useState('');
+  const [residencesMappedToName, setResidencesMappedToName] = useState<{ label: string; value: string; }[]>([]);
+  const { landlord } = useAuth();
+  const [situationReportName, setSituationReportName] = useState('');
+
+  useEffect(() => {
+    const fetchResidences = async () => {
+      if (landlord?.residenceIds) {
```
Comment (mdesautel) [inline app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:367]: you could wrap this in a try-catch, the fetching could go wrong

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+      {toDisplay.map((group, groupIndex) => {
+        const groupName = group[0];
+        const items = group[1];
+
+        if (items.length > 1) {
+          return (
+            <View key={groupIndex} style={situationReportStyles.groupContainer}>
+              {editMode ? (
+                <InputField
+                  value={groupName}
```
Comment (mdesautel) [inline app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:213]: Here you should only put the placeholder as the groupName, not also the value (initially). The user has to manually erase the inital text when wanting to write something. Or is that because you're assuming we will be able to choose from a list of pre-defined groups before adding them?

File: app/screens/landlord/SituationReport/SituationReportScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,425 @@
...
+                height={ButtonDimensions.smallButtonHeight}
+                disabled={false}
+                onPress={async () => {
+
+                  // Hardcoded value for the apartmentId since residence and apartment
+                  const report : SituationReport = {
+                    reportDate: new Date().toISOString(),
+                    residenceId: selectedResidence,
+                    apartmentId: "damH2jH7NRxIVZa0cZgL",
+                    arrivingTenant: JSON.stringify({
```
Comment (mdesautel) [inline app/screens/landlord/SituationReport/SituationReportScreen.tsx:387]: What is the need for stringify here? Can't we just store the data as-is? It seems complex

File: app/__tests__/SituationReport.test.ts [hunk 5]
```diff
@@ -285,3 +285,82 @@ describe('changing the status of an item', () => {
         }).toThrow("Invalid groupIndex, itemIndex or newStatus");
     });
 });
+
+
```
Comment (mdesautel) [inline app/__tests__/SituationReport.test.ts:289]: Wow these test suites are very extensive, good job! Just to make sure, is it normal that this one is named .ts and not .tsx?

**Review by mdesautel — CHANGES_REQUESTED** [review]
Hi, thank you for your PR! The screen looks very good and intuitive, and it has many features to make the user's experience comfortable, it provides a real value to the app! This will allow us to give our app to people so that they can test it in real life and have a meaningful user flow. I mentioned a few things in other comments that you can look into, but overall there are no big issues, as I tested it on my machine and it works smoothly. Also, well done for the tests, the coverage is high and they are super meaningful and complete.
For the 'next step' paragraph in your description, I think it is supposed to be coming very soon, Shin is working on it so we should be able to test everything correctly (without relying on hardcoded values like you have right now, we will have to remember to change that when it is ready).
Let me know if you have questions!

File: app/screens/landlord/SituationReport/SituationReportScreen.tsx (hunk 4 above)
Comment (AdrienDeschenaux) [reply app/screens/landlord/SituationReport/SituationReportScreen.tsx:387]: This is because we decided to have the arriving and leaving tenant names stored in the situation report, since we don't use different types for a person, we need to have the name and the surname. In my opinion there might be situations where the name and the surname could not be easily differentiated (I was thinking of people having maybe multiple surnames, or multiple family names, in a situation where there is strictly more than two strings, then it becomes impossible to really know what could be a surname and what is the family name.)

I agree that maybe a JSON string is not the best option, I proposed earlier in the project to have a type representing a person (since we might need to store personal information of persons that are not necessarly user, which mean we're not certain to have a user document to create a reference to). What could be done there to make it better, could be to create a type to represent a person (name and surname) in the database.

I think it's a good thing that you catched this, but if it's okay for you, I would change this in the task I am assigned to this week (which is a task about polishing everything in those two screens : `SituationReportScreen.tsx` and `SituationReportCreationScreen.tsx` since this is more about `SituationReportCreationScreen.tsx`) as the task related to this PR is more about `SituationReportCreationScreen.tsx (i'm not sure that I changed this screen other than moving it to an other directory during this week)

Let me know your opinion on this!

File: app/__tests__/SituationReport.test.ts (hunk 5 above)
Comment (AdrienDeschenaux) [reply app/__tests__/SituationReport.test.ts:289]: Hey thank you! 

 Yes the fact that one of the file is a `.ts` is totally normal, it is because I'm not testing React Native (`.tsx`) code, but only TypeScript (`.ts`) with those tests, therefore no need for `.tsx` for the test file

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx (hunk 1 above)
Comment (AdrienDeschenaux) [reply app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:598]: Hey! You're absolutely right this should indeed be checked, I'll also update the text reminding the user to select a residence and fill the layout name to make it clear for the user

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx (hunk 2 above)
Comment (AdrienDeschenaux) [reply app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:367]: I agree with you that a try-catch could be a good idea, but the problem is that none of the functions actually return an error, (the authContext should already have handled the error if there were one since the fetching of the landlord informations are fetch during the connection to the app) and the getResidence somehow doesn't have error handling, I think that modifying that in the front-end would be quite messy. We could change this but it would require changes that go beyond the scope of this PR (wrap in a try-catch in the project all the places where getResidence is used), I can however create an issue about this, as a bug in the scrum board if that's ok for you.

Let me know if what I propose is okay for you

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx (hunk 3 above)
Comment (AdrienDeschenaux) [reply app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:212]: Hey here the problem is not with the placeholder, because the placeholder would only be displayed when there is no name, it is because there is a default value for a newly added item, this felt more intuitive for me to put default values because otherwise it looked like it was a bug If there were no initial value. I could switch back to having no initial value in for the group/item name if you prefer

Let me know your opinion on that

**Comment by AdrienDeschenaux** [comment]
Hey Madeline thank you for your review, I made some of the changes and I explained more why I made other implementation choices, I agree on most of the modification you request, however I believe some of these changes might be better addressed in a separate branch. This is because they involve modifications that could impact other parts of the code base, potentially requiring refactoring in other screens.

In summary, I think that your suggestions are important in order to ensure a good behavior of the app as well as a good maintainability, but I suggest tackling them in a separated PR. If you feel strongly that those changes should be part of this PR, I would appreciate if you could share your perspective and I'll be happy to adapt accordingly

Let me know what you think about my suggestions

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx (hunk 2 above)
Comment (mdesautel) [reply app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:366]: Hum yes I noticed too that some methods in the firestore file may be lacking error handling. The getResidence returns null if there is no document, which is an alternative, and as the fetchResidences handles this case with the '?' I think it should actually be safe, what do you think? So you're right, my bad for mentioning this.

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx (hunk 1 above)
Comment (mdesautel) [reply app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx]: Thank you for the fix and the alert message update!

File: app/__tests__/SituationReport.test.ts (hunk 5 above)
Comment (mdesautel) [reply app/__tests__/SituationReport.test.ts:289]: Oh well all good then

File: app/screens/landlord/SituationReport/SituationReportScreen.tsx (hunk 4 above)
Comment (mdesautel) [reply app/screens/landlord/SituationReport/SituationReportScreen.tsx:387]: Oh ok I didn't know you didn't need a review for that screen I just saw it in the changed files😅Yes you can absolutely polish this next week since it is not an error at all!

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx (hunk 2 above)
Comment (AdrienDeschenaux) [reply app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:366]: Yes, if by safe it means it shouldn't crash, I don't think there should be a problem there since the `?` check that there is a value in the field, but regarding the getResidence, If we want to properly handle an error while fetching we might need to improve the firebase functions (which would be a task for an other issue)

Comment (mdesautel) [reply app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:370]: Yes we can add this as an open issue in the scrum board then

File: app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx (hunk 3 above)
Comment (mdesautel) [reply app/screens/landlord/SituationReport/SituationReportCreationScreen.tsx:212]: Thank you for having removed the text and kept the placeholder only, it is way more user-friendly like that:)

**Comment by mdesautel** [comment]
Alright, I've answered all your comments, thank you for the changes you've made. As the CI passes I will approve this and merge your work.

**Review by mdesautel — APPROVED** [review]
